### PR TITLE
Support versionOf in DeltaSharing OSS spark connector

### PR DESCRIPTION
**Pull request by linzhou-db**
Support versionOf in DeltaSharing OSS spark connector

### CODE REVIEW

File: spark/src/main/scala/io/delta/sharing/spark/RemoteDeltaLog.scala [hunk 1]
```diff
@@ -161,13 +167,19 @@ private[sharing] object RemoteDeltaLog {
...
 
   lazy val (metadata, protocol, version) = {
-    val tableMetadata = client.getMetadata(table)
-    (tableMetadata.metadata, tableMetadata.protocol, tableMetadata.version)
+    if (versionOf.isEmpty) {
+      val tableMetadata = client.getMetadata(table)
+      (tableMetadata.metadata, tableMetadata.protocol, tableMetadata.version)
+    } else {
+      val tableFiles = client.getFiles(table, Nil, Some(1L), versionOf)
+      (tableFiles.metadata, tableFiles.protocol, tableFiles.version)
```
Comment (chakankardb) [inline spark/src/main/scala/io/delta/sharing/spark/RemoteDeltaLog.scala:181]: Question: Consider adding a comment that we request only 1 file for efficiency?

File: spark/src/main/scala/io/delta/sharing/spark/RemoteDeltaLog.scala [hunk 2]
```diff
@@ -293,7 +305,7 @@ private[sharing] case class RemoteDeltaFileIndex(
 
   override def refresh(): Unit = {}
 
-  override def sizeInBytes: Long = deltaLog.snapshot.sizeInBytes
```
Comment (chakankardb) [inline spark/src/main/scala/io/delta/sharing/spark/RemoteDeltaLog.scala:296]: I have this change in my PR as well... so perhaps revert from here?
I also got rid of deltaLog arg from the file index.

File: spark/src/test/scala/io/delta/sharing/spark/DeltaSharingSuite.scala [hunk 3]
```diff
@@ -113,6 +113,39 @@ class DeltaSharingSuite extends QueryTest with SharedSparkSession with DeltaShar
     }
   }
 
+  integrationTest("cdf_table_cdf_enabled") {
```
Comment (chakankardb) [inline spark/src/test/scala/io/delta/sharing/spark/DeltaSharingSuite.scala:116]: "cdf_table_cdf_enabled without version" ?

File: spark/src/test/scala/io/delta/sharing/spark/DeltaSharingSuite.scala [hunk 4]
```diff
@@ -113,6 +113,39 @@ class DeltaSharingSuite extends QueryTest with SharedSparkSession with DeltaShar
...
+      Row("2", 2, sqlDate("2020-02-02"))
+    )
+    checkAnswer(spark.read.format("deltaSharing").load(tablePath), expected)
+    withTable("delta_sharing_test") {
+      sql(s"CREATE TABLE delta_sharing_test USING deltaSharing LOCATION '$tablePath'")
+      checkAnswer(sql(s"SELECT * FROM delta_sharing_test"), expected)
+    }
+  }
+
+  integrationTest("cdf_table_cdf_enabled version") {
```
Comment (chakankardb) [inline spark/src/test/scala/io/delta/sharing/spark/DeltaSharingSuite.scala:129]: version -> with valid version?

File: spark/src/main/scala/io/delta/sharing/spark/RemoteDeltaLog.scala (hunk 2 above)
Comment (linzhou-db) [reply spark/src/main/scala/io/delta/sharing/spark/RemoteDeltaLog.scala:308]: This still need it to be correct. I'll wait for your merge first.

File: spark/src/main/scala/io/delta/sharing/spark/RemoteDeltaLog.scala (hunk 1 above)
Comment (linzhou-db) [reply spark/src/main/scala/io/delta/sharing/spark/RemoteDeltaLog.scala:196]: added comment.

File: spark/src/test/scala/io/delta/sharing/spark/DeltaSharingSuite.scala (hunk 3 above)
Comment (linzhou-db) [reply spark/src/test/scala/io/delta/sharing/spark/DeltaSharingSuite.scala]: updated.

File: spark/src/test/scala/io/delta/sharing/spark/DeltaSharingSuite.scala (hunk 4 above)
Comment (linzhou-db) [reply spark/src/test/scala/io/delta/sharing/spark/DeltaSharingSuite.scala]: updated.

File: spark/src/main/scala/io/delta/sharing/spark/RemoteDeltaLog.scala [hunk 5]
```diff
@@ -47,7 +47,13 @@ private[sharing] class RemoteDeltaLog(
 
   @volatile private var currentSnapshot: RemoteSnapshot = new RemoteSnapshot(path, client, table)
 
-  def snapshot: RemoteSnapshot = currentSnapshot
+  def snapshot(versionOf: Option[Long]): RemoteSnapshot = {
```
Comment (chakankardb) [inline spark/src/main/scala/io/delta/sharing/spark/RemoteDeltaLog.scala:50]: default = None?

File: spark/src/main/scala/io/delta/sharing/spark/RemoteDeltaLog.scala [hunk 6]
```diff
@@ -173,13 +180,21 @@ private[sharing] object RemoteDeltaLog {
...
   lazy val (metadata, protocol, version) = {
-    val tableMetadata = client.getMetadata(table)
-    (tableMetadata.metadata, tableMetadata.protocol, tableMetadata.version)
+    if (versionOf.isEmpty) {
+      val tableMetadata = client.getMetadata(table)
+      (tableMetadata.metadata, tableMetadata.protocol, tableMetadata.version)
+    } else {
+      // getMetadata doesn't support the parameter: versionOf
+      // Leveraging getFiles to get the metadata, so setting the limitHint to 1 for efficiency.
+      val tableFiles = client.getFiles(table, Nil, Some(1L), versionOf)
```
Comment (chakankardb) [inline spark/src/main/scala/io/delta/sharing/spark/RemoteDeltaLog.scala:195]: Should we add a method to the client: getMetadata(versionOf) and use it for getting protocol/metadata?

File: spark/src/main/scala/io/delta/sharing/spark/RemoteDeltaLog.scala (hunk 5 above)
Comment (linzhou-db) [reply spark/src/main/scala/io/delta/sharing/spark/RemoteDeltaLog.scala:50]: added.

File: spark/src/main/scala/io/delta/sharing/spark/RemoteDeltaLog.scala (hunk 6 above)
Comment (chakankardb) [reply spark/src/main/scala/io/delta/sharing/spark/RemoteDeltaLog.scala:195]: Perhaps a util method then?

**Review by chakankardb — APPROVED** [review]

**Pull request merged**